src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -225,10 +225,10 @@ public async ValueTask<Variant> GetVariantAsync(string feature, CancellationToke
         /// Gets the assigned variant for a specific feature.
         /// </summary>
         /// <param name="feature">The name of the feature to evaluate.</param>
-        /// <param name="context">An instance of <see cref="TargetingContext"/> used to evaluate which variant the user will be assigned.</param>
+        /// <param name="context">A context providing information that can used to evaluate which variant the user will be assigned.</param>
         /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
         /// <returns>A variant assigned to the user based on the feature's configured allocation.</returns>
-        public async ValueTask<Variant> GetVariantAsync(string feature, TargetingContext context, CancellationToken cancellationToken = default)
+        public async ValueTask<Variant> GetVariantAsync<TContext>(string feature, TContext context, CancellationToken cancellationToken = default)
```

Comment:
Isn't it required for `TContext` to implement `ITargetingContext` for this API to work? If so, what's the point to have `TContext`?

Reply:
While `GetVariantAsync`, the first step is to evaluate the feature flag state according to the feature filter.

If the user uses a contextual feature filter which requires `TContext` for filter evaluation.
`TContext` is not necessarily `ITargetingContext`. The feature manager will still use it to evaluate whether feature flag is enabled or disabled. 
And he can also use two variants through `DefaultWhenEnabled` and `DefaultWhenDisabled` which don't require `TargetingContext` to do the variant allocation.

Previously, we only allow user to pass `TargetingContext` which will make it impossible to use contextual filter. 

How about this?

If the `TContext` doesn't implement `ITargetingContext`, the feature manager will still give it another chance to get targeting context from the targeting context accessor.

Reply:
You are talking about an edge case where `ITargetingContext` may not be necessary, but general speaking, `ITargetingContext` is required for variant assignment. We should enforce the context to implement `ITargetingContext`.

> If the `TContext` doesn't implement `ITargetingContext`, the feature manager will still give it another chance to get targeting context from the targeting context accessor.

No, we shouldn't use ambient context in contextual feature evaluation.

Reply:
> ITargetingContext is required for variant assignment. We should enforce the context to implement ITargetingContext.

`ITargetingContext` is a public interface, we cannot modify it to avoid breaking changes.
If in the future, we want to add new fields in targeting context, we may create a new interface called IAttributeTargetingContext. If we enforce the context to implement ITargetingContext here, we will constrain ourselves.

`TargetingContext` should be sealed. We had an oversight. I created an issue #486 Currently, we are in a uncomfortable spot where we cannot modify anything related to targeting context to avoid potential breaking changes.

Reply:
I also don't think Variants need to be fully tied to targeting. We don't support it yet- but it seems quite reasonable to want variants without wanting user based allocation. 

Making it TContext also aligns it with IsEnabled, which is more consistent. IsEnabled is also able to use Allocation in the same way. So whatever we decide on- the two interfaces should use the same.